index.html

```diff
@@ -257,18 +294,141 @@ <h1>Settings</h1>
+                    });
+                navigator.serviceWorker.addEventListener("message", (event) => {
+                    if (event.data.type === "CACHE_UPDATE") {
+                        savedCDNFiles(event.data.url);
+                        cacheCore();
+                    }
+                });
+            }
+
+            const corePath = "https://cdn.emulatorjs.org/stable/data/cores/";
             loadJSON("/versions", (response) => {
                 if (!response) {
-                    loadJSON("https://cdn.emulatorjs.org/versions.json", resp => loadVersions(resp));
+                    loadJSON("https://cdn.emulatorjs.org/versions.json", (resp) =>
+                        loadVersions(resp)
+                    );
                     document.getElementById("offline-status").textContent = "Offline Status: NOT INSTALLED";
                 } else {
                     loadVersions(response);
                     document.getElementById("offline-status").textContent = "Offline Status: READY";
                 }
             });
+
+            loadJSON(corePath + "cores.json", (response) => {
+                const cacheCores = JSON.parse(localStorage.getItem("cachedCores")) || [];
+                const cores = JSON.parse(response);
+                for (const core of cores) {
+                    const label = document.createElement("label");
+                    const input = document.createElement("input");
+                    input.type = "checkbox";
+                    input.name = "core";
+                    input.value = core.name;
+                    if (cacheCores.includes(core.name)) {
+                        input.checked = true;
+                    }
+                    label.appendChild(input);
+                    label.appendChild(document.createTextNode(core.name));
+                    document.getElementById("cores-form").appendChild(label);
+                }
+            });
+
+            function savedCDNFiles(url) {
+                const cachedFiles =
+                    JSON.parse(localStorage.getItem("cachedCDNFiles")) || [];
+                cachedFiles.push(url);
+                localStorage.setItem("cachedCDNFiles", JSON.stringify(cachedFiles));
+            }
+
+            function cacheCore() {
+                const files = JSON.parse(localStorage.getItem("cachedCDNFiles")) || [];
+                const cacheCores = JSON.parse(localStorage.getItem("cachedCores")) || [];
+                const cores = files.filter((item) => item.endsWith(".json")).map((item) => item.split("/").pop().split(".")[0]);
+                const urlsToCache = [];
+                for (core of cores) {
+                    if (cacheCores.includes(core)) {
+                        continue;
+                    }
+                    urlsToCache.push(corePath + "reports/" + core + ".json");
+                    urlsToCache.push(corePath + core + "-wasm.data");
```

Comment:
Now that I think of it, should we cache all 4 core blobs? Most users are going to only use one or *maybe* two

Reply:
I don't know. Maybe it can check to see if threads are enabled.